### DISCUSSION TITLE
JSON vs JSONL

### DISCUSSION THREAD

**Discussion by jturan** [CMSgov/price-transparency-guide#77] [Q&A] [answered]
Based on the existing discussions I've read on here, the trend seems to be: 1) most developers of these files will go the route of JSON vs XML or YAML, and 2) for larger plans, these files could be massive. That being said, one common issue with massive JSON files are when a system need to load a single file as one large transaction because the data is stored as a single JSON object. [JSONL](https://jsonlines.org/) alleviates this issue by allowing data to be processed one line at a time. Supporting this format might not be necessary if some of the suggested changes are implemented around Group IDs and/or Service Codes, but it might be worth considering by the community to make these files better optimized for consumption.

**Comment by shaselton-usds** [comment, marked as answer]
Hi @jturan 

Yes, definitely makes sense. It is one thing to produce these large files, it is another thing to *read* them. Strategies around consuming these files are really up to those that wish to leverage the information. There are basically two ways to go about this: the whole file is loaded in-memory and processed or there is stream processing to handle larger files.

As it stands, each file is a single, valid JSON object. Creating multiple objects per file (as JSONL does) would technically make the JSON file invalid. That isn't to say that CMS is necessarily against this at this point, but we haven't planned on developing an example supporting this method.

I get the feeling that while a JSONL file would allow for easier consumption, it would greatly increase the size the original JSON file. Sizing would be somewhat negligible when converting CSV files to JSONL, but moving an existing parent-child relationship schema into JSONL does have sizing concerns.

If there is interest in this approach, I would encourage a PR for evaluation.

**Reply by jturan** [reply]
@shaselton-usds good points all around. I didn't think about the top-level JSON values that would essentially be duplicated across every object in the file. I'll think more on schema design, but would also encourage ideas from others on how JSONL could potentially be used in an efficient way.

**Reply by sbenphia** [reply]
Sheesh - @shaselton-usds it is too bad that this thread dropped off a year ago and was shut down so quickly. At least this discussion is what I found when looking for discussion on this issue. Some way to get parts of these files into JSONL would have been a very big help to make these files more readable, but more importantly I think there really should be some guidance on file size for readability as well.  Some networks are indicating these files may be "terabytes" which would essentially make them unreadable (and un-downloadable) for any average data analyst.  Years ago even CMS solved this problem with the SAF and OPPS data by breaking the files up into multiple chunks to be analyzed filtered and loaded (this is a reasonable practice for any large dataset provided by a vendor to someone that they expect to read it).  If these files really are going to be terabytes - it would make sense (and possibly prevent some of this becoming a joke) if there is some guidance in the schema as to how to break the files up into at least gigabyte chunks.  At some point a prohibitive file size seems out of compliance, and does render a file just about as proprietary as putting it in a PDF.  It wouldn't have to be JSONL, but it would make sense to say that the files themselves have to be accessible in digestible sections as a file size, even if there is a little bit of repetition.  Here is what United Health has said in their most recent FAQ:

What is an estimated of the file size? (New 5/17/22)
The files are very large, potentially terabytes. We will be able to provide a more precise
response once the files have been created. 

https://www.uhc.com/content/dam/uhcdotcom/en/HealthReform/PDF/Provisions/reform-external-transparancy-FAQs.pdf

(meant to add humor) - this reminds me of the cartoon put out to show how Russian factories in Communism produced nonsense items (like one large nail to comply with the weight requirement in production) in order to comply with government planning.  https://i.imgur.com/zL6ntxH.jpg
